lib/elixir/src/elixir_erl.erl

```diff
@@ -297,16 +297,20 @@ typespecs_form(Map, Set, Bag, MacroNames) ->
       {Types, Specs, Callbacks, MacroCallbacks, OptionalCallbacks} =
         'Elixir.Kernel.Typespec':translate_typespecs_for_module(Set, Bag),
 
+      DefinedTypePairs = collect_defined_type_pairs(Map, Types),
+      UsedTypePairs = collect_used_type_pairs(Map, DefinedTypePairs, Types, Specs, Callbacks, MacroCallbacks),
+      UsedTypes = filter_used_types(Map, UsedTypePairs, Types),
```

Comment:
Thanks @fertapric! Instead of doing another pass on the types, can't we collect which types are being used as we translate them? So `translate_typespecs_for_module` would return UsedTypes too? This would also reduce the amount of Erlang code we have to write. :)

Reply:
I have other two branches trying to solve this code 😅 

One of them collects the used/unused types during the translation. This was actually my first option, due to the performance benefits it provides, but I found these "issues":

* After the translation, the `caller` env is not available. Errors and warnings require the file and the line, but only the `line` information was there (needed for the `unused_type` warning). I found that I could get the file using `:ets.lookup_element(set, {:cache_env, 0}, 2).file` but I wasn't sure that was the best way to get it. One alternative was to pass the `Map` variable as argument, but I didn't want to change the interface because this is a public function (although it's for internal use only). Other option was to keep the caller when collecting the defined type pairs, something like `%{type_pair => env}`, but I thought the Erlang code was simpler to understand and more intention revealing.
* Collecting the used types during the translation required to pass around some state on any call to `typespec`, and the resulting code was a bit more convoluted. For example, this:

  ```elixir
  defp typespec({:%{}, meta, fields} = map, vars, caller) do
    fields =
      Enum.map(fields, fn
        {k, v} when is_atom(k) ->
          args = [typespec(k, vars, caller), typespec(v, vars, caller)]
          {:type, line(meta), :map_field_exact, args}

        {{:required, meta2, [k]}, v} ->
          args = [typespec(k, vars, caller), typespec(v, vars, caller)]
          {:type, line(meta2), :map_field_exact, args}

        {{:optional, meta2, [k]}, v} ->
          args = [typespec(k, vars, caller), typespec(v, vars, caller)]
          {:type, line(meta2), :map_field_assoc, args}
  ```

  was be changed to:

  ```elixir
  defp typespec({:%{}, meta, fields} = map, vars, caller, types_state) do
    fields =
      Enum.map(fields, fn
        {k, v} when is_atom(k) ->
          {arg1, types_state} = typespec(k, vars, caller, types_state)
          {arg2, types_state} = typespec(v, vars, caller, types_state)
          args = [arg1, arg2]
          {{:type, line(meta), :map_field_exact, args}, types_state}

        {{:required, meta2, [k]}, v} ->
          {arg1, types_state} = typespec(k, vars, caller, types_state)
          {arg2, types_state} = typespec(v, vars, caller, types_state)
          args = [arg1, arg2]
          {{:type, line(meta2), :map_field_exact, args}, types_state}
        
        {{:optional, meta2, [k]}, v} ->
          {arg1, types_state} = typespec(k, vars, caller, types_state)
          {arg2, types_state} = typespec(v, vars, caller, types_state)
          args = [arg1, arg2]
          {{:type, line(meta2), :map_field_assoc, args}, types_state}
  ```
  `typespec/*` functions were doing more than one thing in that case, returning the typespec (its main responsibility) and returning the list of used types.

The other branch does the same thing than the current PR in Elixir, but getting the file using `:ets.lookup_element(set, {:cache_env, 0}, 2).file`.

I'm glad to open a new PR with any of those branches if you think that's the way to go 🙂